### PR TITLE
[NVPTX] Allow MemTransferInst in adjustByValArgAlignment

### PR DESCRIPTION
**Pull request by jsji**
Before b7b28e770c46, AreSupportedUsers will skip
MemTransferInst, it may cause unexpected assertion. https://godbolt.org/z/z5d691fj1
In b7b28e770c46, we start to allow MemTransferInst,
we should allow it in adjustByValArgAlignment too.

### CODE REVIEW

File: llvm/test/CodeGen/NVPTX/lower-byval-args.ll [hunk 1]
```diff
@@ -219,6 +219,20 @@ entry:
   ret void
 }
 
+; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: readwrite)
+define dso_local void @memcpy_from_param_noalign (ptr nocapture noundef writeonly %out, ptr nocapture noundef readonly byval(%struct.S) %s) local_unnamed_addr #0 {
+; COMMON-LABEL: define dso_local void @memcpy_from_param_noalign(
+; COMMON-SAME: ptr nocapture noundef writeonly [[OUT:%.*]], ptr nocapture noundef readonly byval([[STRUCT_S:%.*]]) align 4 [[S:%.*]]) local_unnamed_addr #[[ATTR0]] {
+; COMMON-NEXT:  [[ENTRY:.*:]]
+; COMMON-NEXT:    [[S1:%.*]] = addrspacecast ptr [[S]] to ptr addrspace(101)
+; COMMON-NEXT:    call void @llvm.memcpy.p0.p101.i64(ptr [[OUT]], ptr addrspace(101) [[S1]], i64 16, i1 true)
```
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/lower-byval-args.ll:228]: This appears to be the same output LLVM produces now: https://godbolt.org/z/sbr3hbGsc

```
define dso_local void @memcpy_from_param_noalign(ptr nocapture noundef writeonly %out, ptr nocapture noundef readonly byval(%struct.S) align 4 %s) local_unnamed_addr #0 {
entry:
  %s1 = addrspacecast ptr %s to ptr addrspace(101)
  call void @llvm.memcpy.p0.p101.i64(ptr %out, ptr addrspace(101) %s1, i64 16, i1 true)
  ret void
}
```

Can you update the test so it demonstrates the behavior the patch is intended to fix?

File: llvm/test/CodeGen/NVPTX/lower-byval-args.ll [hunk 2]
```diff
@@ -219,6 +219,20 @@ entry:
   ret void
 }
 
+; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: readwrite)
+define dso_local void @memcpy_from_param_noalign (ptr nocapture noundef writeonly %out, ptr nocapture noundef readonly byval(%struct.S) %s) local_unnamed_addr #0 {
+; COMMON-LABEL: define dso_local void @memcpy_from_param_noalign(
```
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/lower-byval-args.ll:224]: You may also want to make this function a kernel, by adding an appropriate metadata at the end of the file.

File: llvm/test/CodeGen/NVPTX/lower-byval-args.ll (hunk 1 above)
Comment (jsji) [reply llvm/test/CodeGen/NVPTX/lower-byval-args.ll:228]: https://godbolt.org/z/z5d691fj1 Without fix, we are seeing assert, use `assertion trunk` to see the assertion.

File: llvm/test/CodeGen/NVPTX/lower-byval-args.ll (hunk 2 above)
Comment (jsji) [reply llvm/test/CodeGen/NVPTX/lower-byval-args.ll:224]: This is copied from line 207, simply removing `align 4` in param.

Comment (jsji) [reply llvm/test/CodeGen/NVPTX/lower-byval-args.ll:224]: Updated metadata to include new func.

**Review by Artem-B — APPROVED** [review]
LGTM. Thank you for the fix.

File: llvm/test/CodeGen/NVPTX/lower-byval-args.ll (hunk 1 above)
Comment (Artem-B) [reply llvm/test/CodeGen/NVPTX/lower-byval-args.ll]: I see. Thank you!

**Pull request merged**